### PR TITLE
SN-120, SN-119, SN-117 - Access granted when a user-pet and unknown pet detected together, Fixed notification_service to reduce spamming, and Pet needs to be three times before giving access.

### PR DESCRIPTION
**Pull request by deeppatel-11**
SN-120 - Access granted when a user pet and an unknow pet detected together.
SN-117 - Changes to notification to send only one similar notification in one minute.
SN-117 - Pets must be detected 3 times before giving access to them.
- main.py updated for testing 
- FIxed a list bug
- Added test images

### CODE REVIEW

File: WebService/camera.py [hunk 1]
```diff
@@ -13,7 +13,7 @@ class Video(object):
     def __init__(self, queueSize=128, cameraSource=0, timeout=0):
         self.labels = list()
         self.video = CamVideoStream(0)
-        time.sleep(9)
+        time.sleep(4)
```
Comment (MikeG3) [inline WebService/camera.py:16]: I was able to set this timer as low as 1 (second), and it seemed to work fine, and sped up the process, when the camera got disconnected

File: WebService/camera.py [hunk 2]
```diff
@@ -171,23 +172,41 @@ def detecting_dogs_and_cats_body_parts(self, frame):
...
+                                                datetime.datetime.now().strftime(
+                                                    "%Y/%m/%d-%H:%M:%S"), unknown_pet_name)
+            self.pet_detection.draw_bounding_boxes(boxes=self.boxes_coco, index=index,
+                                                   classes=self.coco_classes,
+                                                   class_ids=self.class_ids_coco,
+                                                   confidences=self.confidences_coco, my_img=frame,
+                                                   color=(0, 255, 0), labels=self.labels)
+        #self.detecting_dogs_and_cats_body_parts(frame)
+
+    def recording_bounding_details(self, index):
```
Comment (MikeG3) [inline WebService/camera.py:203]: not a big deal, but suggestion for function name
record_bounding_box_details(self, index)

File: WebService/camera.py [hunk 3]
```diff
@@ -171,23 +172,41 @@ def detecting_dogs_and_cats_body_parts(self, frame):
...
         # This stores the size of each bounding box into a dictionary
-        x, y, w, h = self.boxes_coco[i]
+        x, y, w, h = self.boxes_coco[index]
         # Adding bounding box sizes and proportion of the cat or dog detected
-        self.pet_detection.addingSizeOfBoundingBoxes(str(self.coco_classes[self.class_ids_coco[i]]), w * h)
-        self.pet_detection.addingProportionsOfBoundingBoxes(str(self.coco_classes[self.class_ids_coco[i]]),
-                                                            w, h)
+        self.pet_detection.addingSizeOfBoundingBoxes(str(self.coco_classes[self.class_ids_coco[index]]), w * h)
```
Comment (MikeG3) [inline WebService/camera.py:207]: similar to above suggestion:
addSizeOfBoundingBoxes(...)
addProportionOfBoundingBoxes(...)

File: WebService/camera.py [hunk 4]
```diff
@@ -197,6 +216,8 @@ def get_frame(self):
         # get the height and width
         height, width, _ = frame.shape
         blob = cv2.dnn.blobFromImage(frame, 1 / 255, (416, 416), (0, 0, 0), swapRB=True, crop=False)
+        #blob = cv2.dnn.blobFromImage(frame, 1 / 255, (640, 640), (0, 0, 0), swapRB=True, crop=False)
```
Comment (MikeG3) [inline WebService/camera.py:219]: I think this line can be deleted

File: Pet_Detection/main.py [hunk 5]
```diff
@@ -8,120 +8,106 @@
...
     # if no objects are found then terminate the code.
-    if len(indexes_coco) == 0:
-        print("No object found")
-        sys.exit()
-
-    for i in indexes_coco.flatten():
-        if str(coco_classes[class_ids_coco[i]]) == 'bird' or str(coco_classes[class_ids_coco[i]]) == 'person':
-            pd.draw_bounding_boxes(boxes=boxes_coco, index=i, classes=coco_classes, class_ids=class_ids_coco,
-                                   confidences=confidences_coco, resized_testing_input_image=resized_testing_input_image, color=(0, 0, 255))
-            print(str(coco_classes[class_ids_coco[i]]) + " found near by")
```
Comment (MikeG3) [inline Pet_Detection/main.py:53]: If there is enough time to refactor, Id suggest breaking up each of these lines of codes into separate functions
detect_bird(...) 
if str(coco_classes[class_ids_coco[i]]) == 'bird' or str(coco_classes[class_ids_coco[i]]) == 'person':
            pd.draw_bounding_boxes(boxes=boxes_coco, index=i, classes=coco_classes, class_ids=class_ids_coco,
                                   confidences=confidences_coco, resized_testing_input_image=resized_testing_input_image, color=(0, 0, 255))
            print(str(coco_classes[class_ids_coco[i]]) + " found near by")
 
 detect_person(...)
 ....
detect_mail(..)
etc..

**Review by MikeG3 — COMMENTED** [review]
### _Variables Name
Variable name 'bucket' can be changed to something more descriptive (line 24)
Variable name 'key' can be changed to something more descriptive (line 24)
 
### _Function Name
Theres a few functions that I gave name suggestions, that can be seen in the inline comments
The suggestions are just to remove the grammatical conjugation in the names ("ing")

### _Long Function
Theres a long series of if statements, my suggestion would be to make a function, that analyzes each case, and returns a boolean
*Please see inline comments
Example
if ....
...

Replace with 
def detect_bird()
    if ... 
        return True
    return False
    
### _Code Complexity._
Sonar cloud returned 2 code smells, Ill refer you to Sonar Cloud results

### _PR Checklist
  - [ ] Documentation present for every component and function.
  - [ ] Unit tests are implemented.
  - [x] Implementation matches the corresponding requirements.
  - [x] All commits linked to Jira ticket.
  
### _Overall Remarks._
Great work! Good refactoring and organization of code.
I added an online comment, about lowering the sleep time for activating the camera, it works well on my end, but Im not sure if that will give you issues on your system
As discussed in our team meeting, this all worked well on my end, and we were able to complete 3 of the acceptance tests with this code

**Comment by JChekkouri** [comment]
### Variables Name

The variables names are descriptive enough so it's good.

### Function Name

I don't have any suggestions regarding the function names, they are good.

### Long Function

You could maybe separate some if statements to make it less heavy but otherwise it is good.

### Code Complexity

Sonar cloud returned 3 code smells and 2 bugs.

### PR Checklist

- [ ]  Documentation present for every component and function.


- [ ]  Unit tests are implemented.


- [x]  Implementation matches the corresponding requirements.


- [x]  All commits linked to Jira ticket.


### Overall Remarks

Good work, PR approved.

**Review by JChekkouri — APPROVED** [review]

**Comment by mariemek** [comment]
### Variables Name
the variable names look good and are descriptive enough


### Function Name
Functions look good


### Long Function
Very few long functions and it looks good

### Code Complexity
Sonar cloud returned 3 code smells and 2 bugs.

### PR Checklist

- [ ] Documentation present for every component and function.

- [ ] Unit tests are implemented.



- [x] Implementation matches the corresponding requirements.

 

- [x] All commits linked to Jira ticket.

### Overall Remarks
Good work overall. PR Approved

**Review by mariemek — APPROVED** [review]

**Review by hasan-ahmed — APPROVED** [review]

**Pull request merged**